### PR TITLE
Ability to detect undead ZooKeeper sessions

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

A more correct replacement for https://github.com/ClickHouse/ClickHouse/pull/59735

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 22c97be2a43a683e7f9831533c1fc942004ca00b with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60044/22c97be2a43a683e7f9831533c1fc942004ca00b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60044/22c97be2a43a683e7f9831533c1fc942004ca00b/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Common/ZooKeeper/ZooKeeper.cpp [hunk 1]
```diff
@@ -936,11 +957,50 @@ Coordination::ReconfigResponse ZooKeeper::reconfig(
 }
 
 
+ZooKeeperPtr ZooKeeper::createWithoutKillingPreviousSessions(const ZooKeeperArgs & args_)
+{
+    return std::shared_ptr<ZooKeeper>(new ZooKeeper(args_));
+}
```
Comment (antonio2368) [inline src/Common/ZooKeeper/ZooKeeper.cpp:963]:
```suggestion
template <typename... Args>
ZooKeeperPtr ZooKeeper::createWithoutKillingPreviousSessions(Args &&.... args)
{
    return std::shared_ptr<ZooKeeper>(std::forward<Args>(args)...);
}
```
maybe something like this so we don't have to define this function for each constructor?

**Review by antonio2368 — COMMENTED** [review]
nice

File: src/Common/ZooKeeper/ZooKeeper.cpp [hunk 2]
```diff
@@ -655,21 +666,34 @@ Coordination::Error ZooKeeper::multiImpl(const Coordination::Requests & requests
         auto response = future_result.get();
         Coordination::Error code = response.error;
         responses = response.responses;
+
+        if (check_session_valid)
+        {
+            if (code != Coordination::Error::ZOK && !Coordination::isHardwareError(code) && getFailedOpIndex(code, responses) == requests.size())
+            {
+                impl->finalize(fmt::format("Session was killed: {}", requests.back()->getPath()));
```
Comment (antonio2368) [inline src/Common/ZooKeeper/ZooKeeper.cpp:674]: do we want to continue and return the user error or do we want to transform this into a hardware error?
I know we don't expect such errors really often, but it seems like an internal check that should be retried.

Comment (tavplubix) [reply src/Common/ZooKeeper/ZooKeeper.cpp:674]: Ah, yes, we should convert it to hardware error

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**